tensorboard/uploader/proto/write_service.proto

```diff
@@ -175,77 +218,99 @@ message WriteScalarResponse {
 
 // The blob writing procedure is as follows.  It is intrinsically resumable.
 //  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
-//  * The first of these requests must provide write_blob_first_request_data,
-//    and the subsequent ones must not.
-//  * The first request should send no actual data; it acts as a request for the
-//    current status of the blob, in case it has already been partially or fully
-//    uploaded.
-//  * The first WriteBlobResponse reports the state, size, and crc32c of the
-//    object so far, i.e. taking into account any previously uploaded data.
-//  * The client proceeds to upload from the specified offset (optionally
-//    validating the crc32c up to that point first).
-//  * The caller sends a number of additional WriteBlobRequests, providing the
-//    blob data in an ordered sequence of chunks.
+//  * The caller sends one or more WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
 //  * The caller receives a sequence of WriteBlobResponses in return.  Each
 //    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
 //    the state of the write operation.
+//  * The streams need not operate in 'ping-pong' style-- i.e. the client may
+//    continue to send requests, regardless of whether responses to prior
+//    requests have arrived yet.
+//
+//  * The first request sent over the stream may optionally contain no actual
+//    data. In this case it acts only as a request for the current status of
+//    the blob, in case it has already been partially or fully uploaded.
+//  * The first WriteBlobResponse reports the state, size, and crc32c of the
+//    object so far, i.e. taking into account any previously uploaded data.
+//    This response may imply that any provided data was ignored as redundant.
```

Comment:
I'm not quite sure what this sentence means.  Isn't it an error to provide data at an offset that doesn't correspond to the end of the blob?  So providing redundant data would always be an error, even if it's on the first request.

Reply:
Ah, I see how that was unclear.  I think the current refactoring cleans it all up.